### ISSUE TITLE
Expose remote address and port

### ISSUE THREAD

**Issue by jorams** [orthecreedence/cl-async#135]
Wookie is currently the only Clack backend that doesn't pass `:remote-addr` and `:remote-port` in the environment, which I think is unfortunate.

libuv contains functions called `uv_tcp_getpeername` and `uv_pipe_getpeername`. If I understand correctly, those can be used to retrieve the host and port of the remote end. CL-async doesn't expose that information for a socket and Wookie, by extension, doesn't either.

I'm not sure how to go about implementing this.

**Comment by orthecreedence** [comment]
Agreed, I've let this slide for too long. I'm neck deep in other work right now so this may be open a while, but if I get some free time I'll add this in soon.

One workaround in the meantime is to put nginx/apache as a proxy in front of Wookie and look at the `X-Forwarded-For` header in the incoming request headers. Not ideal, especially if you're just running it internally without a public facing port, but that's the best I can do at the moment =]

**Comment by jorams** [comment]
No rush. Several months ago, working on a little game project, I was looking for this functionality and couldn't find it. That project hasn't seen any activity for a while now, but  when I noticed the discrepancy between the Clack handlers I figured an issue couldn't hurt.

**Comment by cherrymud** [comment]
Any update on this? I'd love to use cl-async for a simple telnet service, but not knowing who is connecting to me makes cl-async a no-go.

**Comment by VaclavSynacek** [comment]
I have a working solution that is just a little bit too ugly and unclear how to merge to cl-async, but it works just fine even without reverse proxy in front of cl-async or other non-lisp workarounds.

When the underneath function is called on `socket` (for example in `:connect-cb`) it sets the `socket-address` to a string containing ip:port of the connected peer.

Questions before this can be turned into pull request:

1. Where should this be put in cl-async? Is it acceptable to turn this on by default? Or should it be a separate function as it is now and let the user call it when he needs the ip and port?
2. I am not sure how to handle the `sa-family` enum. I have seen some cffi tutorial explaining enums in struts, but I am not sure this means I need to send pull request also to cl-libuv to change the type of `sa-family` to enum. A temporary ugly hack is the check for `2` meaning IPv4.
3. Should both ip address and port go to slot `address`? Or add slot `port` to `socket`? And shouldn't they be renamed to `peer-address` and `peer-port`?

```lisp
(defpackage :socket-address
  (:use :cl :cl-async)
  (:export :setf-socket-address))

(in-package :socket-address)

(defun setf-socket-address (socket)
  (let
    ((c-socket (socket-c socket))
     (max-ip-string-length 50))
    (cffi:with-foreign-string (c-ip-string
                                (format nil "~v@{~A~:*~}"
                                        max-ip-string-length
                                        #\Space))
      (cffi:with-foreign-objects
        ((c-addr '(:struct libuv:sockaddr))
         (c-addr-length :int))
        (setf (cffi:mem-aref c-addr-length :int)
              (cffi:foreign-type-size '(:struct libuv:sockaddr)))
        (libuv:uv-tcp-getpeername c-socket c-addr c-addr-length)
        (case (cffi:foreign-slot-value
                c-addr
                '(:struct libuv:sockaddr)
                'libuv:sa-family)
          (2 ;IPv4 FIXME proper ENUM decoding 
             (progn
               (libuv:uv-ip-4-name
                 c-addr c-ip-string max-ip-string-length)
               (let
                 ((ip (cffi:foreign-string-to-lisp c-ip-string))
                  (port (cffi:foreign-slot-value
                          c-addr
                          '(:struct libuv:sockaddr-in)
                          'libuv:sin-port)))
                 (setf (cl-async::socket-address socket)
                       (format nil "~a:~a" ip port)))))
          (t ;IPv6 FIXME proper ENUM decoding 
             (progn
               (libuv:uv-ip-6-name
                 c-addr c-ip-string max-ip-string-length)
               (let
                 ((ip (cffi:foreign-string-to-lisp c-ip-string))
                  (port (cffi:foreign-slot-value
                          c-addr
                          '(:struct libuv:sockaddr-in6)
                          'libuv:sin6-port)))
                 (setf (cl-async::socket-address socket)
                       (format nil "~a:~a" ip port))))))))))
```

PS: solution inspired by [this](https://github.com/HaxeFoundation/hashlink/issues/226#issuecomment-457948333).
